packages/tooltip/src/Tooltip.tsx

```diff
@@ -0,0 +1,307 @@
+}
+
+/**
+ * # Tooltip
+ *
+ * Tooltip component
+ *
+ * ```
+<Tooltip
+  align='top'
+  justify='start'
+  trigger={<button>trigger</button>}
+  triggerEvent='hover'
+  variant='light'
+>
+  I am an uncontrolled Tooltip!
+</Tooltip>
+```
+ * @param props.children Content to appear inside of Tooltip.
+ * @param props.open Boolean to describe whether or not Tooltip is open.
+ * @param props.setOpen Callback to change the open state of the Tooltip.
+ * @param props.variant Whether the Tooltip should be `dark` or `light`.
+ * @param props.className Classname applied to Tooltip.
+ * @param props.align Alignment of Tooltip relative to trigger: `top`, `bottom`, `left`, `right`.
+ * @param props.justify Justification of Tooltip relative to trigger: `start`, `middle`, `end`.
+ * @param props.trigger Trigger element can be ReactNode or function.
+ * @param props.triggerEvent Whether the Tooltip should be triggered by a `click` or `hover`.
+ * @param props.id id given to Tooltip content.
+ */
+function Tooltip({
+  open: controlledOpen,
+  setOpen: controlledSetOpen,
+  className,
+  children,
+  trigger,
+  variant = Variant.Light,
+  triggerEvent = TriggerEvent.Hover,
+  align = 'top',
+  justify = 'start',
+  id,
+  shouldClose,
+  ...rest
+}: TooltipProps) {
+  const isControlled = typeof controlledOpen === 'boolean';
+  const [uncontrolledOpen, uncontrolledSetOpen] = useState(false);
+  const open = isControlled ? controlledOpen : uncontrolledOpen;
+  // typescript is not recognizing isControlled checks that controlledSetOpen exists
+  const setOpen = isControlled
+    ? (controlledSetOpen as Function)
+    : uncontrolledSetOpen;
+
+  const tooltipRef = useRef<HTMLDivElement>(null);
+
+  const tooltipId = useMemo(
+    () => id || `tooltip-${Math.floor(Math.random() * Math.floor(10))}`,
+    [id],
+  );
+
+  const createTriggerProps = (
+    triggerEvent: TriggerEvent,
+    triggerProps?: any,
+  ) => {
+    if (triggerEvent === TriggerEvent.Hover) {
+      return {
+        onMouseEnter: debounce(() => {
+          setOpen((curr: boolean) => !curr);
+        }, 35),
+        onMouseLeave: debounce(handleClose, 35),
+        onFocus: () => setOpen(true),
+        onBlur: handleClose,
+      };
+    }
+
+    if (triggerProps && triggerProps.onClick) {
+      return {
+        onClick: (e: MouseEvent) => {
+          // ensure that we don't close the tooltip when content inside tooltip is clicked
+          if (e.target !== tooltipRef.current) {
+            setOpen((curr: boolean) => !curr);
+            triggerProps.onClick();
+          }
+        },
+      };
+    }
+
+    return {
+      onClick: (e: MouseEvent) => {
+        // ensure that we don't close the tooltip when content inside tooltip is clicked
+        if (e.target !== tooltipRef.current) {
+          setOpen((curr: boolean) => !curr);
+        }
+      },
+    };
+  };
+
+  const handleClose = () => {
+    if (typeof shouldClose !== 'function' || shouldClose()) {
```

Comment:
This logic is wrong, right?

```suggestion
    if (typeof shouldClose === 'function' && shouldClose()) {
```

Before we change this code can we write a test that covers this case?

Reply:
We're checking that shouldClose either doesn't exist so we can go ahead and close the tooltip automatically, or that shouldClose() returns true and we can still proceed with closing the tooltip. Happy to write a test though :)

Reply:
If `shouldClose` is set to a non-function, like a string, then the logic breaks.

Reply:
We're defaulting to handling the open/close behavior ourselves. The shouldClose() is just an optional hook into this by allowing a consumer to pass a callback function that can return true/false to determine whether or not something should close when we think it should. Happy to add a test, but I don't think the logic is broken.